util.c

```diff
@@ -201,3 +201,75 @@ int r_mkdir(char *path)
 	}
 	return 0;
 }
+
+void construct_argv(char **argv, unsigned int len, ...)
+{
+	unsigned int i;
+	va_list args;
+
+	va_start(args, len);
+	for (i = 0; i < len; ++i)
+		argv[i] = va_arg(args, char *);
+	va_end(args);
+	if (argv[len-1] != NULL)
+		error(EXIT_FAILURE, 0, "argv not NULL terminated");
+}
+
+static void xclose(int fd)
+{
+	if (fd >= 0)
+		close(fd);
+}
+
+spawn_t spawn(const char *cmd, char *const argv[], unsigned int flags)
+{
+	pid_t pid;
+	spawn_t status = { -1, -1, -1 };
+	int pfd_read[2] = { -1, -1 };
+	int pfd_write[2] = { -1, -1 };
+	const bool r = flags & X_READ;
+	const bool w = flags & X_WRITE;
+
+	if (cmd == NULL || argv == NULL || flags == 0)
+		return status;
+
+	if (r && pipe(pfd_read) < 0) {
+		error(0, errno, "pipe: %s", cmd);
+		return status;
+	}
+
+	if (w && pipe(pfd_write) < 0) {
+		xclose(pfd_read[0]);
+		xclose(pfd_read[1]);
```

Comment:
pfd_read will have valid fds if and only if w and r are true at this point. Seems simple enough to check rather than having a special wrapper.

Reply:
I personally think not having to check in multiple places was cleaner. But no strong opinion, can change it to explicitly check on spot.

Reply:
Done in https://github.com/nsxiv/nsxiv/pull/211/commits/d11138ce9bc0e3c35000083f9a858b15c486ea3e . If you still think this is cleaner then feel free to mark this conversation as resolved, I'm fine either ways @TAAPArthur .